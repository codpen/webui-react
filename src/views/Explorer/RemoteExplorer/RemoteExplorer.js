import React from 'react';
import {Card, CardBody, Container} from "reactstrap";
import FilesView from "../FilesView/FilesView";
import {addColonAtLast} from "../../../utils/Tools";
import {connect} from "react-redux";
import * as PropTypes from 'prop-types';
import {
	changePath,
	changeRemoteName,
	changeRemotePath,
	createPath,
	navigateBack,
	navigateFwd,
	navigateUp
} from "../../../actions/explorerStateActions";
import FileOperations from "../../Base/FileOperations/FileOperations";
import {PROP_CURRENT_PATH, PROP_FS_INFO} from "../../../utils/RclonePropTypes";
import ErrorBoundary from "../../../ErrorHandling/ErrorBoundary";
import RemotesList from "../RemotesList";


class RemoteExplorer extends React.Component {
	constructor(props) {
		super(props);
		this.state = {
			remoteNameTemp: ""
		};

		this.updateRemoteName = this.updateRemoteName.bind(this);
		this.updateRemotePath = this.updateRemotePath.bind(this);
	}


	updateRemoteName(remoteName) {
		this.setState({remoteNameTemp: remoteName});
	}

	updateRemotePath(newRemotePath, IsDir, IsBucket) {
		const {remoteName} = this.props.currentPath;

		let updateRemoteName = "";
		let updateRemotePath = "";

		if (IsBucket) {
			updateRemoteName = addColonAtLast(remoteName) + newRemotePath;
			updateRemotePath = "";

		} else if (IsDir) {
			updateRemoteName = remoteName;
			updateRemotePath = newRemotePath;
		}
		this.props.changePath(this.props.containerID, updateRemoteName, updateRemotePath);
	}

	render() {


		const {remoteName} = this.props.currentPath;
		const {containerID, className} = this.props;

		const isValidPath = remoteName && remoteName !== "";

		return (
			<ErrorBoundary>

				<Card className={className}>
					<CardBody>
						<Container fluid={true}>

							{isValidPath ? (
								<>
									<FileOperations containerID={containerID}/>
									<FilesView containerID={containerID}/>
								</>
							) : (
								<RemotesList
									remoteName={remoteName}
									containerID={containerID}
								/>
							)}
						</Container>

					</CardBody>
				</Card>


			</ErrorBoundary>
		);


	}

}


const propTypes = {

	containerID: PropTypes.string.isRequired,
	createPath: PropTypes.func.isRequired,
	currentPath: PROP_CURRENT_PATH,
	fsInfo: PROP_FS_INFO,
	hasError: PropTypes.bool,
	distractionFreeMode: PropTypes.bool.isRequired,
	className: PropTypes.string

};

const defaultProps = {};

const mapStateToProps = (state, ownProps) => {

	const currentPath = state.explorer.currentPaths[ownProps.containerID];
	let fsInfo = {};

	const {remoteName} = currentPath;

	if (currentPath && state.remote.configs) {

		const tempRemoteName = remoteName.split(':')[0];
		if (state.remote.configs[tempRemoteName])

			fsInfo = state.remote.configs[tempRemoteName];
	}
	return {
		configs: state.remote.configs,
		hasError: state.remote.hasError,
		error: state.remote.error,
		currentPath,
		fsInfo,
	}

};

RemoteExplorer.propTypes = propTypes;
RemoteExplorer.defaultProps = defaultProps;

export default connect(
	mapStateToProps,
	{
		createPath, changePath,
		changeRemoteName, changeRemotePath, navigateUp,
		navigateBack, navigateFwd
	}
)(RemoteExplorer);
