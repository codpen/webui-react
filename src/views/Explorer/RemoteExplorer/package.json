{
  "name": "RemoteExplorer",
  "version": "0.0.0",
  "private": true,
  "main": "RemoteExplorer.js"
}
