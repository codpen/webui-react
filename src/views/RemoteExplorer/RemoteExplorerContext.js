import React from "react";

const RemoteExplorerContext = React.createContext();

export default RemoteExplorerContext;

