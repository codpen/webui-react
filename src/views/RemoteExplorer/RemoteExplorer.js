import React from 'react';
import "../../utils/Global";
import {Button, Card, CardBody, CardHeader, Col, Form, Row} from "reactstrap";
import RemotesList from "../RemotesList";
import FilesView from "../FilesView/FilesView";
import ScrollableDiv from "../Base/ScrollableDiv/ScrollableDiv";
import RemoteExplorerContext from "./RemoteExplorerContext";
import {addColonAtLast} from "../../utils/Tools";
import {connect} from "react-redux";
import {getConfigForRemote} from "../../actions/explorerActions";
import PropTypes from 'prop-types';
import {
    changePath,
    changeRemoteName,
    changeRemotePath,
    createPath,
    navigateBack,
    navigateFwd,
    navigateUp
} from "../../actions/explorerStateActions";


class RemoteExplorer extends React.Component {
    constructor(props) {
        super(props);
        this.state = {

            remoteName: "",
            remotePath: "",
            remoteNameTemp: ""
        };


        this.updateRemoteName = this.updateRemoteName.bind(this);
        this.updateRemotePath = this.updateRemotePath.bind(this);
        this.buttonUpPressed = this.buttonUpPressed.bind(this);

    }

    getFsInfo(remoteName) {
        console.log("config for", remoteName);

        if (!this.props.configs[remoteName])
            this.props.getConfigForRemote(remoteName);

    }

    updateRemoteName(remoteName) {

        this.setState({remoteNameTemp: remoteName});

    }

    openRemote = () => {
        const {remoteNameTemp} = this.state;
        // backStack.empty();
        // backStack.push({remoteName: remoteNameTemp, remotePath: ""});
        // this.setState(backStack.peek());

        this.props.changeRemoteName(this.props.containerID, remoteNameTemp, "");

        this.getFsInfo(remoteNameTemp);

    };

    updateRemotePath(newRemotePath, IsDir, IsBucket) {
        // const {backStack} = this.props;
        const {remoteName, remotePath} = this.props.currentPath;
        // const curRemotePath = backStack.peek().remotePath;

        let updateRemoteName = "";
        let updateRemotePath = "";

        console.log(remoteName, remotePath, newRemotePath)

        if (IsBucket) {
            updateRemoteName = addColonAtLast(remoteName) + newRemotePath;
            updateRemotePath = "";
            // backStack.push({remoteName: addColonAtLast(backStack.peek().remoteName) + remotePath, remotePath: ""});

        } else if (IsDir) {
            updateRemoteName = remoteName;
            updateRemotePath = newRemotePath;
            // backStack.push({remoteName: backStack.peek().remoteName, remotePath: remotePath});
        }
        this.props.changePath(this.props.containerID, updateRemoteName, updateRemotePath);
    }

    buttonUpPressed() {
        this.props.navigateUp(this.props.containerID);
        // const {backStack} = this.state;
        // if (backStack.getLength() > 1) {
        //     backStack.pop();
        //
        //     this.setState(backStack.peek());
        // }
    }

    componentDidUpdate(prevProps, prevState, snapshot) {
        if (prevState.componentShouldUpdate) {
            this.setState({componentShouldUpdate: false});

        }
    }


    render() {

        const currentPath = this.props.currentPath;
        if (!currentPath) return null;
        console.log(currentPath);


        const {remoteName, remotePath} = currentPath;

        const fsInfo = this.props.configs[remoteName];
        console.log(remotePath, remoteName, fsInfo);
        return (
            <RemoteExplorerContext.Provider
                value={{remoteName: remoteName, remotePath: remotePath, fsInfo: fsInfo}}>
                {/*Render remotes array*/}

                <Card>
                    <CardHeader>Remotes</CardHeader>
                    <CardBody>
                        <Form onSubmit={() => this.openRemote()}>
                            <Row>
                                <Col xs={12} sm={8} lg={6}>

                                    <RemotesList updateRemoteNameHandle={this.updateRemoteName}
                                                 remoteName={remoteName}/>
                                </Col>
                                <Col xs={12} sm={2} lg={2}>

                                    <Button className={"btn-lg"} color="success"
                                            type="submit">Open</Button>
                                </Col>
                            </Row>
                        </Form>
                    </CardBody>
                </Card>

                {/*Render the files in the selected remote*/}
                <Card>
                    <CardHeader>
                        Files: {remoteName}
                    </CardHeader>
                    <CardBody>
                        <ScrollableDiv height={"700px"}>
                            {/*<Row className={"mr-0 ml-0"}>*/}
                            <FilesView remoteName={remoteName} remotePath={remotePath}
                                       updateRemotePathHandle={this.updateRemotePath}
                                       upButtonHandle={this.buttonUpPressed}
                                       componentShouldUpdate={this.state.componentShouldUpdate}
                            />
                            {/*</Row>*/}
                        </ScrollableDiv>
                    </CardBody>
                </Card>
            </RemoteExplorerContext.Provider>
        );

    }

}


const propTypes = {

    containerID: PropTypes.string.isRequired,
    createPath: PropTypes.func.isRequired,
};

const defaultProps = {};

const mapStateToProps = (state, ownProps) => ({
    configs: state.remote.configs,
    hasError: state.remote.hasError,
    error: state.remote.error,
    // backStack: state.explorer.backStacks[ownProps.containerID],
    currentPath: state.explorer.currentPaths[ownProps.containerID],
    // remotePath: state.explorer.currentPaths[ownProps.containerID].remotePath
});

RemoteExplorer.propTypes = propTypes;
RemoteExplorer.defaultProps = defaultProps;

export default connect(
    mapStateToProps,
    {
        getConfigForRemote, createPath, changePath,
        changeRemoteName, changeRemotePath, navigateUp,
        navigateBack, navigateFwd
    }
)(RemoteExplorer);
