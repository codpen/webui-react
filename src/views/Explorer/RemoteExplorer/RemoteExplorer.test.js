import React from "react";
import {Provider} from "react-redux";
import {testStore} from "../../../../Utils";
import {shallow} from "enzyme";
import RemoteExplorer from "./RemoteExplorer";

const setUp = (props = {}) => {
    const component = shallow(
        <Provider store={testStore()}>
            <RemoteExplorer/>
        </Provider>
    );
    return component;
}

describe('Bandwidth Status Card', function () {

    let wrapper;
    beforeEach(() => {
        const props = {};
        wrapper = setUp(props)
    });

    it('should render without crashing', function () {
        expect(wrapper).toHaveLength(1)
    });
});

