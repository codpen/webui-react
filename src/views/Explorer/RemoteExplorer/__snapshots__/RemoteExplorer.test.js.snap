// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Remote Explorer renders should match snapshot 1`] = `
<ErrorBoundary>
  <Card
    tag="div"
  >
    <CardHeader
      tag="div"
    >
      Remotes
    </CardHeader>
    <CardBody
      tag="div"
    >
      <Connect(RemotesList)
        containerID="0"
        remoteName="local"
      />
    </CardBody>
  </Card>
  <Card
    tag="div"
  >
    <CardHeader
      tag="div"
    >
      <Connect(FileOperations)
        containerID="0"
      />
    </CardHeader>
    <CardBody
      tag="div"
    >
      <Connect(DropTarget(FilesView))
        containerID="0"
      />
    </CardBody>
  </Card>
</ErrorBoundary>
`;
